activerecord/lib/active_record/connection_adapters/postgresql/schema_creation.rb

```diff
@@ -57,6 +57,19 @@ def add_column_options!(sql, options)
             if options[:collation]
               sql << " COLLATE \"#{options[:collation]}\""
             end
+
+            if as = options[:as]
+              sql << " GENERATED ALWAYS AS (#{as})"
+
+              if options[:stored]
+                sql << " STORED"
+              else
+                raise ArgumentError, <<~MSG
+                  PostgreSQL currently does not support VIRTUAL (not persisted) generated columns.
+                  Specify 'stored: true' option for '#{options[:column].name}'
+                MSG
```

Comment:
If we omit `STORED`, does it result in an error? That may be preferable, considering future versions may support non-stored virtual columns.

Reply:
As of PG 12 and PG 13, it is required:
> GENERATED ALWAYS AS ( generation_expr ) STORED
> ...
> The keyword STORED is required to signify that the column will be computed on write and will be stored on disk.
Source: https://www.postgresql.org/docs/12/sql-createtable.html
Source: https://www.postgresql.org/docs/13/sql-createtable.html

It looks like their plan is to require the keyword `STORED` and in future versions, require the keyword `VIRTUAL`

Reply:
Can we implement this with an expectation of compatibility with future PG versions?

Reply:
Predicting the future is always a bit tricky :)

In regard to your original suggestion, the thought is to let the exception happen deeper down the stack (with the `pg` gem) and let that error be the guide - as opposed to creating this error. Is that right?

If that is correct, I think the trade-off is:
- As written
  - Con: More brittle in handling postgres updates
  - Pro: Gives the developer a precise fix in the error message (`Specify 'stored: true' option for ...`)

- Letting pg gem raise the error
  - Pro: handles future postgresql updates more accurately/flexibly
  - Con: Leaves it up to the developer to google the error to find the fix

@fanfilmu - WDYT?

Reply:
Giving the developer a precise fix in the error message is definitely a nice feature.